File: dotnet/Lexy/Lexy.Compiler/Language/Tables/Table.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Language.Types;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Tables
{
    public class Table : RootNode
    {
        public TableName Name { get; } = new TableName();
        public TableHeader Header { get; private set; }
        public IList<TableRow> Rows { get; } = new List<TableRow>();

        public override string NodeName => Name.Value;

        public const string RowName = "Row";

        private Table(string name, SourceReference reference) : base(reference)
        {
            Name.ParseName(name);
        }

        internal static Table Parse(NodeName name, SourceReference reference)
        {
            return new Table(name.Name, reference);
        }

        public override IParsableNode Parse(IParserContext context)
        {
            var line = context.CurrentLine;
            if (line.IsEmpty()) return this;

            if (line.IsComment())
            {
                throw new InvalidOperationException("No comments expected. Comment should be parsed by Document only.");
            }

            if (IsFirstLine())
            {
                Header = TableHeader.Parse(context);
            }
            else
            {
                Rows.Add(TableRow.Parse(context));
            }

            return this;
        }

        private bool IsFirstLine()
        {
            return Header == null;
        }

        public override IEnumerable<INode> GetChildren()
        {
            if (Header != null)
            {
                yield return Header;
            }

            foreach (var row in Rows)
            {
                yield return row;
            }
        }

        protected override void Validate(IValidationContext context)
        {
        }

        public override void ValidateTree(IValidationContext context)
        {
            using (context.CreateVariableScope())
            {
                base.ValidateTree(context);
            }
        }

        public ComplexType GetRowType(IValidationContext context)
        {
            var members = Header.Columns
                .Select(column => new ComplexTypeMember(column.Name, column.Type.CreateVariableType(context)))
                .ToList();

            return new ComplexType(Name.Value, ComplexTypeSource.TableRow, members);
        }
    }
}

File: typescipt/src/language/tables/table.ts


export class Table extends RootNode {
   public const string RowName = `Row`;
   public TableName Name new(): =;
   public TableHeader Header { get; private set; }
   public Array<TableRow> Rows = list<TableRow>(): new;

   public override string NodeName => Name.Value;

   private Table(string name, SourceReference reference) {
     super(reference);
     Name.ParseName(name);
   }

   internal static parse(name: NodeName, reference: SourceReference): Table {
     return new Table(name.Name, reference);
   }

   public override parse(context: IParseLineContext): IParsableNode {
     if (IsFirstLine()) {
       Header = TableHeader.parse(context);
     }
     else {
       Rows.Add(TableRow.parse(context));
     }

     return this;
   }

   private isFirstLine(): boolean {
     return Header == null;
   }

   public override getChildren(): Array<INode> {
     if (Header != null) yield return Header;

     foreach (let row in Rows) yield return row;
   }

   protected override validate(context: IValidationContext): void {
   }

   public override validateTree(context: IValidationContext): void {
     using (context.CreateVariableScope()) {
       base.ValidateTree(context);
     }
   }

   public getRowType(context: IValidationContext): ComplexType {
     let members = Header.Columns
       .Select(column => new ComplexTypeMember(column.Name, column.Type.createVariableType(context)))
       .ToList();

     return new ComplexType(Name.Value, ComplexTypeSource.TableRow, members);
   }
}
